### ISSUE TITLE
Parsing GLnexus pVCF

### ISSUE THREAD

**Issue by aderzelle** [knausb/vcfR#152]
Hello,

I was wondering if you had ever tried to parse a GLnexus pvcf? https://github.com/dnanexus-rnd/GLnexus

A priori, it seems vcfR reads it without complaining but afterwards I am not able to query the sample column in the dat table, it seems to ignore them
This is how the pVCF looks like

```
##contig=<ID=chromosome_34,length=44>
##contig=<ID=chromosome_35,length=43>
##bcftools_viewVersion=1.9+htslib-1.9
##bcftools_viewCommand=view ALL.nexus.bcf; Date=Thu Feb 20 09:32:56 2020
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	ancestor	D2A1	D2B3	D3A1	D4A3	D5B3	H2A3	H2C3	H4A4	H4C2	H5A3
chromosome_1	6340194	chromosome_1_6340194_G_C	G	C	43	.	AF=0.454545;AQ=43	GT:DP:AD:GQ:PL:RNC	0/1:301:164,132:31:31,0,40:..	./.:.:.:.:0,0,0:MM	0/1:43:26,17:37:37,0,56:..	0/1:45:23,22:34:35,0,38:..	0/1:55:24,31:34:33,0,53:..	0/1:82:41,41:26:26,0,42:..	0/1:44:33,11:27:27,0,49:..	0/1:134:82,52:34:34,0,48:..	0/1:49:34,15:34:33,0,59:..	0/1:211:118,92:43:43,0,56:..	0/1:201:102,99:41:41,0,56:..
```
and this is what vcfR sees 

```
pvcf = read.vcfR("/home/alessandro/Data/18-02-2020_DeepVariant/ARCcestor_mosquito_model/ALL.nexus.vcf.gz", verbose = FALSE)
ext = vcfR2tidy(pvcf, single_frame = TRUE, info_types = TRUE, format_types = TRUE)

```

```
 ext
$dat
# A tibble: 13,075,040 x 19
   CHROM            POS ID                       REF   ALT    QUAL FILTER AF       AQ    AC       AN Indiv        gt_GT gt_RNC gt_DP gt_AD   gt_GQ gt_PL   gt_GT_alleles
   <chr>          <int> <chr>                    <chr> <chr> <dbl> <chr>  <chr>    <chr> <chr> <int> <chr>        <chr> <chr>  <int> <chr>   <int> <chr>   <chr>        
 1 chromosome_1 6340194 chromosome_1_6340194_G_C G     C        43 NA     0.454545 43    NA       NA ARC_ancestor 0/1   ..       301 164,132    31 31,0,40 G/C          
 2 chromosome_1 6340194 chromosome_1_6340194_G_C G     C        43 NA     0.454545 43    NA       NA D2A1         NA    MM        NA NA         NA 0,0,0   .            
 3 chromosome_1 6340194 chromosome_1_6340194_G_C G     C        43 NA     0.454545 43    NA       NA D2B3         0/1   ..        43 26,17      37 37,0,56 G/C          
 4 chromosome_1 6340194 chromosome_1_6340194_G_C G     C        43 NA     0.454545 43    NA       NA D3A1         0/1   ..        45 23,22      34 35,0,38 G/C          
 5 chromosome_1 6340194 chromosome_1_6340194_G_C G     C        43 NA     0.454545 43    NA       NA D4A3         0/1   ..        55 24,31      34 33,0,53 G/C          
 6 chromosome_1 6340194 chromosome_1_6340194_G_C G     C        43 NA     0.454545 43    NA       NA D5B3         0/1   ..        82 41,41      26 26,0,42 G/C          
 7 chromosome_1 6340194 chromosome_1_6340194_G_C G     C        43 NA     0.454545 43    NA       NA H2A3         0/1   ..        44 33,11      27 27,0,49 G/C          
 8 chromosome_1 6340194 chromosome_1_6340194_G_C G     C        43 NA     0.454545 43    NA       NA H2C3         0/1   ..       134 82,52      34 34,0,48 G/C          
 9 chromosome_1 6340194 chromosome_1_6340194_G_C G     C        43 NA     0.454545 43    NA       NA H4A4         0/1   ..        49 34,15      34 33,0,59 G/C          
10 chromosome_1 6340194 chromosome_1_6340194_G_C G     C        43 NA     0.454545 43    NA       NA H4C2         0/1   ..       211 118,92     43 43,0,56 G/C     
```     

Any suggestion? Thank you

EDIT: the pVCF follows VCF specifications VCFv4.2

**Comment by knausb** [comment]
Hi @aderzelle, technically this does not appear to be a valid VCF file because the first line does not specify what version of the specification the file adheres to. Based on your example vcfR appear to be working fine. I do not understand what your goal is. You state you're interested in 'parsing' the data which is vague. Do you have a specific goal? You reference a 'sample column' but this does not exist in the VCF specification. Can you create a [minimal reproducible example](https://knausb.github.io/vcfR_documentation/reporting_issue.html) where there is something specific I can help you with? Thanks!
Brian

**Comment by aderzelle** [comment]
Oh yes sorry, here is the complete header 
```
##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##GLnexusVersion=v1.2.5-0-g3b5f961
##GLnexusConfigName=DeepVariantWGS
##GLnexusConfigCRC32C=2857227159
##GLnexusConfig={unifier_config: {drop_filtered: false, min_allele_copy_number: 1, min_AQ1: 10, min_AQ2: 10, min_GQ: 0, max_alleles_per_site: 32, monoallelic_sites_for_lost_alleles: true, preference: common}, genotyper_config: {revise_genotypes: true, min_assumed_allele_frequency: 9.99999975e-05, required_dp: 0, allow_partial_data: true, allele_dp_format: AD, ref_dp_format: MIN_DP, output_residuals: false, more_PL: true, squeeze: false, trim_uncalled_alleles: true, output_format: BCF, liftover_fields: [{orig_names: [MIN_DP, DP], name: DP, description: "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">", type: int, number: basic, default_type: missing, count: 1, combi_method: min, ignore_non_variants: true}, {orig_names: [AD], name: AD, description: "##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">", type: int, number: alleles, default_type: zero, count: 0, combi_method: min, ignore_non_variants: false}, {orig_names: [GQ], name: GQ, description: "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">", type: int, number: basic, default_type: missing, count: 1, combi_method: min, ignore_non_variants: true}, {orig_names: [PL], name: PL, description: "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Phred-scaled genotype Likelihoods\">", type: int, number: genotype, default_type: missing, count: 0, combi_method: missing, ignore_non_variants: true}]}}
##INFO=<ID=AF,Number=A,Type=Float,Description="Allele Frequency estimate for each alternate allele">
##INFO=<ID=AQ,Number=A,Type=Integer,Description="Allele Quality score reflecting evidence for each alternate allele (Phred scale)">
##INFO=<ID=AC,Number=A,Type=Integer,Description="Allele count in genotypes">
##INFO=<ID=AN,Number=1,Type=Integer,Description="Total number of alleles in called genotypes">
##FILTER=<ID=MONOALLELIC,Description="Site represents one ALT allele in a region with multiple variants that could not be unified into non-overlapping multi-allelic sites">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=RNC,Number=2,Type=Character,Description="Reason for No Call in GT: . = n/a, M = Missing data, P = Partial data, I = gVCF input site is non-called, D = insufficient Depth of coverage, - = unrepresentable overlapping deletion, L = Lost/unrepresentable allele (other than deletion), U = multiple Unphased variants present, O = multiple Overlapping variants present, 1 = site is Monoallelic, no assertion about presence of REF or ALT allele">
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Approximate read depth (reads with MQ=255 or with bad mates are filtered)">
##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Allelic depths for the ref and alt alleles in the order listed">
##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">
##FORMAT=<ID=PL,Number=G,Type=Integer,Description="Phred-scaled genotype Likelihoods">
##contig=<ID=chromosome_1,length=20333293>
##contig=<ID=chromosome_2,length=17705893>
##contig=<ID=chromosome_3,length=16800825>
##contig=<ID=chromosome_4,length=16096788>
##contig=<ID=chromosome_5,length=15031336>
##contig=<ID=chromosome_6,length=13873466>
##contig=<ID=chromosome_7,length=19194>
##contig=<ID=chromosome_8,length=14922>
##contig=<ID=chromosome_9,length=4214>
##contig=<ID=chromosome_10,length=3769>
##contig=<ID=chromosome_11,length=3600>
##contig=<ID=chromosome_12,length=2517>
##contig=<ID=chromosome_13,length=1651>
##contig=<ID=chromosome_14,length=1479>
##contig=<ID=chromosome_15,length=1152>
##contig=<ID=chromosome_16,length=1138>
##contig=<ID=chromosome_17,length=1000>
##contig=<ID=chromosome_18,length=971>
##contig=<ID=chromosome_19,length=967>
##contig=<ID=chromosome_20,length=967>
##contig=<ID=chromosome_21,length=940>
##contig=<ID=chromosome_22,length=757>
##contig=<ID=chromosome_23,length=545>
##contig=<ID=chromosome_24,length=465>
##contig=<ID=chromosome_25,length=390>
##contig=<ID=chromosome_26,length=306>
##contig=<ID=chromosome_27,length=197>
##contig=<ID=chromosome_28,length=169>
##contig=<ID=chromosome_29,length=148>
##contig=<ID=chromosome_30,length=147>
##contig=<ID=chromosome_31,length=68>
##contig=<ID=chromosome_32,length=66>
##contig=<ID=chromosome_33,length=65>
##contig=<ID=chromosome_34,length=44>
##contig=<ID=chromosome_35,length=43>
##bcftools_viewVersion=1.9+htslib-1.9
##bcftools_viewCommand=view ALL.nexus.bcf; Date=Thu Feb 20 09:32:56 2020
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	ancestor	D2A1	D2B3D3A1	D4A3	D5B3	H2A3	H2C3	H4A4	H4C2	H5A3

```
Each column after the "FORMAT" column represents a different biological sample. I am interested in singletons: SNPs that are present in ONE single sample and NOT in other samples 

The VCF looks like so 

```
chromosome_1    6340194 chromosome_1_6340194_G_C        G       C       43      .       AF=0.454545;AQ=43       GT:DP:AD:GQ:PL:RNC      0/1:301:164,132:31:31,0,40:..   ./.:.:.:.:0,0,0:MM      0/1:43:26,17:37:37,0,56:..      0/1:45:23,22:34:35,0,38:..      0/1:55:24,31:34:33,0,53:..      0/1:82:41,41:26:26,0,42:..      0/1:44:33,11:27:27,0,49:..      0/1:134:82,52:34:34,0,48:..     0/1:49:34,15:34:33,0,59:..      0/1:211:118,92:43:43,0,56:..    0/1:201:102,99:41:41,0,56:..
chromosome_1    6340269 chromosome_1_6340269_A_T        A       T       55      .       AF=0.5;AQ=55    GT:DP:AD:GQ:PL:RNC      0/1:292:159,131:43:43,0,55:..   0/1:42:17,25:48:49,0,54:..      0/1:43:25,18:39:39,0,65:..      0/1:51:29,22:55:55,0,62:..      0/1:47:19,28:50:50,0,63:..      0/1:85:43,42:46:46,0,54:..      0/1:46:30,16:37:37,0,66:..      0/1:128:61,67:52:52,0,69:..     0/1:50:28,22:44:43,0,66:..      0/1:200:120,78:48:48,0,62:..    0/1:215:113,102:42:41,0,57:..
chromosome_1    6340334 chromosome_1_6340334_C_T        C       T       60      .       AF=0.5;AQ=60    GT:DP:AD:GQ:PL:RNC      0/1:297:158,135:38:38,0,44:..   0/1:51:21,30:55:60,0,55:..      0/1:46:25,21:44:46,0,48:..      0/1:56:32,24:46:45,0,64:..      0/1:47:17,30:51:58,0,51:..      0/1:88:47,41:50:51,0,57:..      0/1:50:34,16:43:42,0,70:..      0/1:145:62,82:50:50,0,57:..     0/1:53:27,26:54:54,0,67:..      0/1:197:111,85:42:41,0,60:..    0/1:209:99,110:46:46,0,54:..
chromosome_1    6340544 chromosome_1_6340544_T_C        T       C       48      .       AF=0.5;AQ=48    GT:DP:AD:GQ:PL:RNC      0/1:310:155,154:33:33,0,44:..   0/1:55:29,26:49:48,0,67:..      0/1:56:29,27:36:35,0,62:..      0/1:55:28,27:45:44,0,71:..      0/1:65:29,36:31:31,0,52:..      0/1:75:33,42:38:37,0,60:..      0/1:62:35,27:38:37,0,63:..      0/1:166:70,96:37:37,0,58:..     0/1:59:43,16:29:28,0,66:..      0/1:175:90,85:38:37,0,59:..     0/1:220:122,97:38:38,0,52:..

```
Here all the samples have a 0/1 genotype so those variants do not interest me. So I am looking for lines where one sample has GT=0/1 or GT=0/2 or GT=./1 or GT=./2 and, I think, for example 0/0 in all other samples. That's not exactly space rocket coding. 

(I understand that joint calling like this might not be absolutely appropriate as it might bias the call towards variants that are shared).

**Comment by knausb** [comment]
How about this?

``` r
library(vcfR)
#> 
#>    *****       ***   vcfR   ***       *****
#>    This is vcfR 1.10.0 
#>      browseVignettes('vcfR') # Documentation
#>      citation('vcfR') # Citation
#>    *****       *****      *****       *****
# reprex looks in a temp dir so we need to tell it where the file is.
setwd("~/Desktop/sandbox/aderzelle/")
pvcf <- read.vcfR("~/Desktop/sandbox/aderzelle/my_pVCF.vcf")
#> Scanning file to determine attributes.
#> File attributes:
#>   meta lines: 54
#>   header_line: 55
#>   variant count: 1
#>   column count: 20
#> Meta line 54 read in.
#> All meta lines processed.
#> gt matrix initialized.
#> Character matrix gt created.
#>   Character matrix gt rows: 1
#>   Character matrix gt cols: 20
#>   skip: 0
#>   nrows: 1
#>   row_num: 0
#> Processed variant: 1
#> All variants processed
pvcf
#> ***** Object of Class vcfR *****
#> 11 samples
#> 1 CHROMs
#> 1 variants
#> Object size: 0 Mb
#> 0 percent missing data
#> *****        *****         *****
# Fabricate more than one variant
pvcf@fix <- rbind(pvcf@fix, pvcf@fix, pvcf@fix)
pvcf@gt  <- rbind(pvcf@gt, pvcf@gt, pvcf@gt)
# ID must be unique
pvcf@fix[, "ID"] <- paste(pvcf@fix[, "ID"], letters[1:3], sep="_")
pvcf
#> ***** Object of Class vcfR *****
#> 11 samples
#> 1 CHROMs
#> 3 variants
#> Object size: 0 Mb
#> 0 percent missing data
#> *****        *****         *****
# We appear to only be interested in genotypes.
gt <- extract.gt(pvcf)
# Set all homozygous.
gt[3, ] <- "0/0"
# Add singleton.
gt[3, 2] <- "0/1"
# Create helper function.
is.singleton <- function(x){
  x <- unlist(strsplit(x, "/", fixed = TRUE))
  if( sum(na.omit(x) == "0") == 1 ){ return(TRUE) }
  if( sum(na.omit(x) == "1") == 1 ){ return(TRUE) }
  if( sum(na.omit(x) == "2") == 1 ){ return(TRUE) }
  return(FALSE)
}
apply(gt, MARGIN = 1, FUN = is.singleton)
#> chromosome_1_6340194_G_C_a chromosome_1_6340194_G_C_b 
#>                      FALSE                      FALSE 
#> chromosome_1_6340194_G_C_c 
#>                       TRUE
```

<sup>Created on 2020-02-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Note that this does not handle phased genotypes and throws away missing alleles. Let me know if this is a solution for you.

**Comment by aderzelle** [comment]
Woaw thanks, that's pretty neat. The only limitation is that after applying the function we lose the information of from which sample the variant comes, so when I query 
```
singletons = apply(gt, MARGIN = 1, FUN = is.singleton)
fin = grep("TRUE", singletons, value = TRUE)
 head(fin)
chromosome_1_6351082_T_C chromosome_1_6360550_C_A chromosome_1_6388950_G_T chromosome_1_6392301_T_C chromosome_1_6392383_A_T chromosome_1_6400633_C_T 
                  "TRUE"                   "TRUE"                   "TRUE"                   "TRUE"                   "TRUE"                   "TRUE" 
```
I don't readily get the information.  But it's possible to be retrieved of course, so far I am doing this:

```
GT = as.data.table(gt, keep.rownames = TRUE)
#transforming into data.table for easy query
GT[GT$rn == "chromosome_1_6351082_T_C"]
                         rn ARC_ancestor D2A1 D2B3 D3A1 D4A3 D5B3 H2A3 H2C3 H4A4 H4C2 H5A3
1: chromosome_1_6351082_T_C         <NA> <NA> <NA> <NA> <NA> <NA> <NA> <NA>  0/1 <NA> <NA>
```
And now I know that this SNP is unique to the H4A4 sample. 
In fact, if we think in term of a function, it should return the position (which is given by first column of gt) and then the sample ID. So in my example

`chromosome_1_6351082_T_C  H4A4`

I would like to add that this kind of query are common for people working in non model organisms, doing experimental evolution, etc ... so it might be worth thinking of adding it in the package.

**Comment by knausb** [comment]
If we are to consider this as a part of vcfR I think we need a solution that is more robust to various scenarios that a user may throw at it. And perhaps I did not explain well how I thought you could use this. Here's another try.

``` r
library(vcfR)
#> 
#>    *****       ***   vcfR   ***       *****
#>    This is vcfR 1.10.0 
#>      browseVignettes('vcfR') # Documentation
#>      citation('vcfR') # Citation
#>    *****       *****      *****       *****
data("vcfR_test")
vcfR_test
#> ***** Object of Class vcfR *****
#> 3 samples
#> 1 CHROMs
#> 5 variants
#> Object size: 0 Mb
#> 0 percent missing data
#> *****        *****         *****

is.singleton <- function(x, max_allele = 2){
  if( !inherits(x, "vcfR") ){
    stop( 
      paste("Expecting an object of class vcfr, received instead", class(x)) 
          )
  }
  
  is.sing <- function(y){
    y <- unlist(strsplit(y, "/", fixed = TRUE))
    y <- unlist(strsplit(y, "|", fixed = TRUE))
    result <- FALSE
    for(i in 0:max_allele){
      if( sum( na.omit(y) == as.character(i) ) == 1 ){ 
        result <- TRUE 
      }
    }
    return(result)
  }

  gt <- extract.gt(x)
  apply(gt, MARGIN = 1, FUN = is.sing)
}

my_singletons <- is.singleton(vcfR_test)
my_singletons
#>  rs6054257   20_17330  rs6040355 20_1230237  microsat1 
#>      FALSE       TRUE      FALSE      FALSE       TRUE

sing_vcf <- vcfR_test[my_singletons, ]
getFIX(sing_vcf)
#>      CHROM POS       ID          REF   ALT      QUAL FILTER
#> [1,] "20"  "17330"   NA          "T"   "A"      "3"  "q10" 
#> [2,] "20"  "1234567" "microsat1" "GTC" "G,GTCT" "50" "PASS"
sing_vcf@gt
#>      FORMAT        NA00001          NA00002        NA00003   
#> [1,] "GT:GQ:DP:HQ" "0|0:49:3:58,50" "0|1:3:5:65,3" "0/0:41:3"
#> [2,] "GT:GQ:DP"    "0/1:35:4"       "0/2:17:2"     "1/1:40:3"
```

<sup>Created on 2020-02-21 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Note that a variant may include more than one singleton. I think this function could better support the sort of info in VCF files and not just your specific issue. Would this be a solution? We could also consider a function that would call this one.

**Comment by aderzelle** [comment]
Thank you for the detailed explanation. What exactly do you mean by 

> Note that a variant may include more than one singleton.
if the vcf is properly formatted there should only be a single variant per line right?

**Comment by aderzelle** [comment]
Actually, something like bcftools --private --samples is what would be needed.

**Comment by knausb** [comment]
Hmm, I think you're confusing me now. I'm afraid I see your example as poor because there is only one genotype and it is heterozygous. This means it would be private by multiple definitions. Can we try again? Perhaps it would help if we tried some definitions.

**Singleton** I see a singleton allele as one that occurs once over all of the samples.
**Private to a sample** This allele may occur in only one sample. For example, ```2/2``` represents an allele that occurs twice. But if it only occurs in one sample it is private to that sample.
**Private to a population** An allele may occur multiple times but is only present in one population.

Does one of these match the scenario you're looking for? If not, can you provide an attempt?

You are correct that in VCF data there should be one variant per line or row. But each variant may have multiple alleles. In the example data I presented above the variant with the ID ```microsat1``` has two alternate alleles, this mean that one of them has the potential to fit any of the above definitions.  To my knowledge there is no limit to the number of alternate alleles in the VCF specification. SNP data may be limited to A, C, G, T, but microsats and structural variants may be unlimited.

I do not use bcftools (because I use vcfR) but it seems to me that the command you've provided will find alleles private to a population. Is this correct? Thanks!

**Comment by aderzelle** [comment]
Hello, yes sure. 

I agree with your definition, then let's say I am looking for private alleles to a population. I am working with a clonal microorganism so each of my population should behave as a "single individual" at the genetic level (in my experimental setup at least).

The command I gave does indeed find private alleles. However, I reasoned that in order to compare populations, it only makes sense to do that at sites that can be genotyped across all the samples. Therefore in the example I gave, I would reject that site, because as it is "unknown" in the other samples, I don't know if my allele is private because it's not in the other populations, or if it's private because it occurs at a site that is very hard to sequence, or because it's an artefact. 

The vcf is generated with GLnexus and the MM means that there is just no information about that sites in all my populations but one. 

In other words, I am looking for sites that are private to a single sample AND that are genotyped in all other samples. However, this might not be a standard demand. Hence I am not sure if it is worth that you implement a dedicated function in vcfR.
